packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -35,19 +35,28 @@ import "./ISignalService.sol";
 contract SignalService is AuthorizableContract, ISignalService {
     using SafeCast for uint256;
 
+    /// @dev Meaning, for this chainId A, signal service contract B, in given snapshot world state
+    /// root 'C', then here is a storage proof for that signal service.
+    mapping(
+        uint256 signalServiceChainId
+            => mapping(
+                address signalServiceAddress
+                    => mapping(bytes32 worldStateRoot => bytes32 signalStorageRoot)
+            )
+    ) public verifiedStorageRoots;
```

Comment:
How is this cache used then?

Reply:
What do you mean ?
It saves at LibTrieProof the heavy on-chain calculation of the storage root.